lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp

```diff
@@ -800,6 +802,32 @@ void rewriteSlice(SetVector<Value> &slice, DenseMap<Value, Attribute> &layout,
         int numIndVars = newForOp.getNumInductionVars();
         mapping.map(loopBody.getArgument(m.first + numIndVars),
                     loopBody.getArgument(m.second + numIndVars));
+        LLVM_DEBUG({
+          DBGS() << "mapping forOp "
+                 << loopBody.getArgument(m.first + numIndVars) << " to "
+                 << loopBody.getArgument(m.second + numIndVars) << '\n';
+        });
+        // Check to see if there are existing convertOps on the old Argument
+        Value oldRes = forOp.getResult(m.first);
+        Value newRes = newForOp.getResult(m.second);
+        for (Operation *user : oldRes.getUsers()) {
+          if (auto convertOp = dyn_cast<ConvertLayoutOp>(user)) {
```

Comment:
This is probably too narrow in scope as it only works when the convert operation comes immediately after the for loop. I tried going down the other path of enabling rematerialization through for loops:
https://github.com/openai/triton/compare/main...peterbell10:triton:rematerialize-through-for

However ran into an issue where a tensor may be rematerialized by multiple different users, which creates duplicates and the pass relies on CSE to clean them up. However, there is no canonicalization pattern to CSE two loop variables (and in general I think that would be a non-trivial pass to write). This meant you would end up recomputing everything in the loop body multiple times, even with the same layout :/

One idea could be to keep a record of which tensors have equivalent values, but different layouts and add a shortcut in the rematerialization phase that looks these up and avoids creating duplicates in the first place. It would probably require a lot of bookkeeping though which was putting me off.

Reply:
I actually saw this pattern a while back from an internal benchmark but didn't spend time looking into it. It is great to know that you are working on the other path :]
In your patch your change is after
```
if (currentValue.getDefiningOp<scf::ForOp>())
      return failure();
```
I wonder how it actually has an effect if we already bail out with ForOp.

> However, there is no canonicalization pattern to CSE two loop variables (and in general I think that would be a non-trivial pattern to write).

Yes, the rematerialization depends on CSE to clean up duplications. Do you have an example IR for this?

> One idea could be to keep a record of which tensors have equivalent values, but different layouts and add a shortcut in the rematerialization phase that looks these up and avoids creating duplicates in the first place. It would probably require a lot of bookkeeping though which was putting me off.

This is more like my approach where we keep a record of values and rematerialized values so we can use the existing rematerialized values instead of remating the slice again.

Reply:
> I wonder how it actually has an effect if we already bail out with ForOp.

Sorry, that was an old commit. I've pushed the working version which removes the bail out and also handles `ForOp` in `rewriteSlice`.

> Yes, the rematerialization depends on CSE to clean up duplications. Do you have an example IR for this?

See the `rematerialize_if_inside_loop` test I added previously:

https://github.com/openai/triton/blob/740b985bcd86a91ce0fbba7a997025d45c19e38b/test/TritonGPU/combine.mlir#L2208

```mlir
// -----
#blocked = #triton_gpu.blocked<{sizePerThread = [2], threadsPerWarp = [32], warpsPerCTA = [1], order = [0]}>
module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 1 : i32, "triton_gpu.threads-per-warp" = 32 : i32} {
  tt.func public @rematerialize_if_inside_loop() -> (tensor<32xf32, #blocked>, tensor<32xf32, #blocked>) {
    %c4096_i32 = arith.constant 4096 : i32
    %c32_i32 = arith.constant 32 : i32
    %c0_i32 = arith.constant 0 : i32
    %cst = arith.constant dense<2.000000e+00> : tensor<32xf32, #blocked>
    %cst_0 = arith.constant dense<1.000000e+00> : tensor<32xf32, #blocked>
    %0:3 = scf.for %arg0 = %c0_i32 to %c4096_i32 step %c32_i32 iter_args(%arg1 = %cst, %arg2 = %cst_0, %arg3 = %cst_0) -> (tensor<32xf32, #blocked>, tensor<32xf32, #blocked>, tensor<32xf32, #blocked>)  : i32 {
      %1 = arith.cmpi eq, %arg0, %c0_i32 : i32
      %2:3 = scf.if %1 -> (tensor<32xf32, #blocked>, tensor<32xf32, #blocked>, tensor<32xf32, #blocked>) {
        scf.yield %cst, %cst_0, %cst_0 : tensor<32xf32, #blocked>, tensor<32xf32, #blocked>, tensor<32xf32, #blocked>
      } else {
        %3 = arith.addf %arg2, %cst_0 : tensor<32xf32, #blocked>
        %4 = arith.addf %arg3, %cst_0 : tensor<32xf32, #blocked>
        %5 = arith.mulf %arg1, %3 : tensor<32xf32, #blocked>
        scf.yield %5, %3, %4 : tensor<32xf32, #blocked>, tensor<32xf32, #blocked>, tensor<32xf32, #blocked>
      }
      scf.yield %2#0, %2#1, %2#2 : tensor<32xf32, #blocked>, tensor<32xf32, #blocked>, tensor<32xf32, #blocked>
    }
    tt.return %0#2, %0#0 : tensor<32xf32, #blocked>, tensor<32xf32, #blocked>
  }
}
```

Where `%arg1` from the original IR has been rematerialized by both the convert layout inside the loop and by rematerializing through the loop. 

> This is more like my approach where we keep a record of values and rematerialized values so we can use the existing rematerialized values instead of remating the slice again.

IIUC your PR as written simply delays the erasure of a dead `ConvertLayoutOp` because it may already be in the queue to be rematerialized. I'm talking about for every mlir value, even those not resulting from a `ConvertLayoutOp`, because you could have for example a chain of operations that get rematerialized starting from different points along the chain. e.g. 

```mlir
%1 = arith.addf %0, %cst_0 : tensor<32xf32, #blocked>
%2 = arith.addf %1, %cst_0 : tensor<32xf32, #blocked>

%3 = triton_gpu.convert_layout %1: tensor<32xf32, #blocked1>
%4 = triton_gpu.convert_layout %2: tensor<32xf32, #blocked1>
```

In this case `%4` would rematerialize both `%1` and `%2`, even though `%1` was already rematerialized for `%3`. We would potentially need to remember this information between multiple different runs of the remove layout conversion pass.

Reply:
I suppose for that specific test case a more complete form of dead argument elimination would be enough, since the value `%0#1` is never used anywhere. In general though you could have multiple users outside the for loop which are rematerialized independently so would need CSE.

Reply:
> between multiple different runs of the remove layout conversion pass

I feel we can have a class for rematerialization similar to LayoutPropagation, we can keep track of the remat'ed values together with the original values, before rematerializing, we can check the remat data structure first. We can erase the old values at end of the pass.

But this will not be persistent across different runs of the pass. Do we really need to keep it across different runs?

Reply:
> I suppose for that specific test case a more complete form of dead argument elimination would be enough, since the value `%0#1` is never used anywhere. In general though you could have multiple users outside the for loop which are rematerialized independently so would need CSE.

Yes the rematerialized argument/output is not used, we should teach DAE to remove it. But after removing it, the convert_layout on the result outside of the loop will still exist, for which remat may be more beneficial.

Reply:
This pass is called 3 times so in theory you could get up to 3x duplication of loop variables. I suppose it would still be an improvement over the current state which is pretty much unbounded duplication, so yeah it might not be necessary for now.